File: android/app/src/main/java/org/manchestermakerspace/portal/MakerspaceNfcPlugin.java
package org.manchestermakerspace.portal;

import android.nfc.NfcAdapter;
import android.nfc.NdefMessage;
import android.nfc.NdefRecord;
import android.nfc.Tag;
import android.nfc.tech.Ndef;
import android.net.Uri;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Locale;

@CapacitorPlugin(name = "MakerspaceNfc")
public class MakerspaceNfcPlugin extends Plugin {
    private volatile int generation = 0;
    private volatile boolean scanning = false;
    private String activeSession = "";
    private NfcAdapter adapter() { return NfcAdapter.getDefaultAdapter(getContext()); }

    @PluginMethod public void capabilities(PluginCall call) {
        NfcAdapter nfc = adapter();
        JSObject result = new JSObject();
        result.put("supported", nfc != null); result.put("enabled", nfc != null && nfc.isEnabled());
        call.resolve(result);
    }

    @PluginMethod public void start(PluginCall call) {
        final String mode = call.getString("mode", "ndef");
        if (!Arrays.asList("ndef", "inspect", "enroll").contains(mode)) { call.reject("Invalid scan mode"); return; }
        getActivity().runOnUiThread(() -> {
            NfcAdapter nfc = adapter();
            if (nfc == null) { call.reject("This device has no NFC reader."); return; }
            if (!nfc.isEnabled()) { call.reject("Enable NFC in Android Settings, then scan again."); return; }
            final int session = ++generation;
            activeSession = call.getString("sessionId", "");
            scanning = true;
            int flags = NfcAdapter.FLAG_READER_NFC_A | NfcAdapter.FLAG_READER_NFC_B |
                NfcAdapter.FLAG_READER_NFC_F | NfcAdapter.FLAG_READER_NFC_V;
            if (mode.equals("enroll")) flags |= NfcAdapter.FLAG_READER_SKIP_NDEF_CHECK;
            try {
                nfc.enableReaderMode(getActivity(), tag -> read(tag, mode, session), flags, null);
                call.resolve();
            } catch (Exception error) { scanning = false; call.reject("Could not start NFC scanning.", error); }
        });
    }

    private void read(Tag tag, String mode, int session) {
        if (!scanning || generation != session) return;
        JSObject result = new JSObject();
        JSArray urls = new JSArray(), texts = new JSArray(), unsupported = new JSArray();
        if (!mode.equals("ndef")) {
            StringBuilder uid = new StringBuilder();
            for (byte b : tag.getId()) uid.append(String.format(Locale.ROOT, "%02X", b & 0xff));
            if (uid.length() > 0) result.put("uid", uid.toString());
        }
        if (!mode.equals("enroll")) {
            Ndef ndef = Ndef.get(tag);
            if (ndef != null) {
                try {
                    ndef.connect();
                    NdefMessage message = ndef.getNdefMessage();
                    if (message != null) decode(message, urls, texts, unsupported, 0);
                } catch (Exception error) { result.put("warning", "NDEF could not be read. Hold the card steady and try again."); }
                finally { try { ndef.close(); } catch (Exception ignored) {} }
            }
        }
        result.put("urls", urls); result.put("texts", texts); result.put("unsupported", unsupported);
        getActivity().runOnUiThread(() -> {
            if (!scanning || generation != session) return;
            stopReader(); notifyListeners("tag", result);
        });
    }

    private void decode(NdefMessage message, JSArray urls, JSArray texts, JSArray unsupported, int depth) throws Exception {
        if (depth > 4 || message.toByteArray().length > 32768 || message.getRecords().length > 64) throw new Exception("Oversized NDEF");
        boolean hasPoster = false;
        for (NdefRecord record : message.getRecords()) if (record.getTnf() == NdefRecord.TNF_WELL_KNOWN && Arrays.equals(record.getType(), NdefRecord.RTD_SMART_POSTER)) hasPoster = true;
        for (NdefRecord record : message.getRecords()) {
            byte[] payload = record.getPayload();
            if (record.getTnf() == NdefRecord.TNF_WELL_KNOWN && Arrays.equals(record.getType(), NdefRecord.RTD_SMART_POSTER)) {
                decode(new NdefMessage(payload), urls, texts, unsupported, depth + 1);
            } else if (record.getTnf() == NdefRecord.TNF_WELL_KNOWN && Arrays.equals(record.getType(), NdefRecord.RTD_TEXT)) {
                if (payload.length == 0) continue;
                int languageLength = payload[0] & 63;
                if (languageLength + 1 > payload.length) throw new Exception("Invalid NDEF text");
                Charset encoding = (payload[0] & 128) != 0 ? StandardCharsets.UTF_16 : StandardCharsets.UTF_8;
                texts.put(new String(payload, languageLength + 1, payload.length - languageLength - 1, encoding));
            } else {
                Uri uri = record.toUri();
                if (uri != null && (record.getTnf() == NdefRecord.TNF_ABSOLUTE_URI || Arrays.equals(record.getType(), NdefRecord.RTD_URI))) {
                    if (!hasPoster) urls.put(uri.toString());
                } else if (record.getTnf() != NdefRecord.TNF_EMPTY) unsupported.put("TNF " + record.getTnf());
            }
        }
    }

    private void stopReader() {
        scanning = false; generation++;
        NfcAdapter nfc = adapter();
        if (nfc != null) nfc.disableReaderMode(getActivity());
    }
    @PluginMethod public void stop(PluginCall call) { getActivity().runOnUiThread(() -> {
        if (activeSession.equals(call.getString("sessionId", ""))) stopReader();
        call.resolve();
    }); }
    @Override protected void handleOnPause() { stopReader(); }
    @Override protected void handleOnDestroy() { stopReader(); }
}

File: native/ios/MakerspaceNfcPlugin.swift
import Capacitor
import CoreNFC

@objc(MakerspaceNfcPlugin)
public class MakerspaceNfcPlugin: CAPPlugin, CAPBridgedPlugin, NFCTagReaderSessionDelegate {
    public let identifier = "MakerspaceNfcPlugin"
    public let jsName = "MakerspaceNfc"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "capabilities", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "start", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stop", returnType: CAPPluginReturnPromise)
    ]
    private var session: NFCTagReaderSession?
    private var mode = "ndef"
    private var activeSession = ""
    private var backgroundObserver: NSObjectProtocol?
    override public func load() {
        backgroundObserver = NotificationCenter.default.addObserver(forName: UIApplication.didEnterBackgroundNotification, object: nil, queue: .main) { [weak self] _ in self?.stopSession() }
    }
    deinit { if let observer = backgroundObserver { NotificationCenter.default.removeObserver(observer) } }

    @objc func capabilities(_ call: CAPPluginCall) {
        call.resolve(["supported": NFCTagReaderSession.readingAvailable, "enabled": NFCTagReaderSession.readingAvailable])
    }
    @objc func start(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            guard NFCTagReaderSession.readingAvailable else { call.reject("NFC is unavailable on this device."); return }
            self.stopSession()
            self.activeSession = call.getString("sessionId") ?? ""
            self.mode = call.getString("mode") ?? "ndef"
            guard ["ndef", "inspect", "enroll"].contains(self.mode) else { call.reject("Invalid scan mode"); return }
            // ISO14443 covers supported MIFARE tags; Classic is not supported by Core NFC.
            self.session = NFCTagReaderSession(pollingOption: [.iso14443, .iso15693], delegate: self, queue: .main)
            self.session?.alertMessage = "Hold your iPhone near the card."
            self.session?.begin()
            call.resolve()
        }
    }
    @objc func stop(_ call: CAPPluginCall) { DispatchQueue.main.async {
        if self.activeSession == call.getString("sessionId") { self.stopSession() }; call.resolve()
    } }
    private func stopSession() { let previous = session; session = nil; previous?.invalidate() }
    public func tagReaderSessionDidBecomeActive(_ session: NFCTagReaderSession) {}
    public func tagReaderSession(_ reader: NFCTagReaderSession, didInvalidateWithError error: Error) {
        guard session === reader else { return }
        session = nil
        notifyListeners("scanError", data: ["message": error.localizedDescription])
    }
    public func tagReaderSession(_ reader: NFCTagReaderSession, didDetect tags: [NFCTag]) {
        guard session === reader else { return }
        guard tags.count == 1, let tag = tags.first else { reader.alertMessage = "Present one card at a time."; reader.restartPolling(); return }
        reader.connect(to: tag) { error in
            guard self.session === reader else { return }
            if let error = error { reader.invalidate(errorMessage: error.localizedDescription); return }
            let uid: Data
            let ndef: NFCNDEFTag
            switch tag {
            case .miFare(let value): uid = value.identifier; ndef = value
            case .iso7816(let value): uid = value.identifier; ndef = value
            case .iso15693(let value): uid = value.identifier; ndef = value
            case .feliCa(let value): uid = value.currentIDm; ndef = value
            @unknown default: reader.invalidate(errorMessage: "Unsupported NFC card."); return
            }
            var result: [String: Any] = ["urls": [String](), "texts": [String](), "unsupported": [String]()]
            if self.mode != "ndef" { result["uid"] = uid.map { String(format: "%02X", $0) }.joined() }
            let finish = { (data: [String: Any]) in
                DispatchQueue.main.async {
                    guard self.session === reader else { return }
                    self.stopSession(); self.notifyListeners("tag", data: data)
                }
            }
            if self.mode == "enroll" { finish(result); return }
            ndef.queryNDEFStatus { status, _, error in
                guard self.session === reader else { return }
                guard error == nil, status != .notSupported else { finish(result); return }
                ndef.readNDEF { message, error in
                    guard self.session === reader else { return }
                    if let message = message {
                        do {
                            let decoded = try self.decode(message, depth: 0)
                            result.merge(decoded) { _, new in new }
                        } catch { result["warning"] = "NDEF message is malformed or too large." }
                    } else if error != nil { result["warning"] = "NDEF could not be read. Try again." }
                    finish(result)
                }
            }
        }
    }
    private func decode(_ message: NFCNDEFMessage, depth: Int) throws -> [String: Any] {
        guard depth <= 4, message.records.count <= 64, message.length <= 32768 else { throw NSError(domain: "NDEF", code: 1) }
        var urls = [String](), texts = [String](), unsupported = [String]()
        let hasPoster = message.records.contains { $0.typeNameFormat == .nfcWellKnown && $0.type == Data("Sp".utf8) }
        for record in message.records {
            if record.typeNameFormat == .nfcWellKnown && record.type == Data("Sp".utf8) {
                guard let nested = NFCNDEFMessage(data: record.payload) else { throw NSError(domain: "NDEF", code: 2) }
                let content = try decode(nested, depth: depth + 1)
                urls += content["urls"] as? [String] ?? []; texts += content["texts"] as? [String] ?? []
                unsupported += content["unsupported"] as? [String] ?? []
            } else if let url = record.wellKnownTypeURIPayload() {
                if !hasPoster { urls.append(url.absoluteString) }
            } else if record.typeNameFormat == .absoluteURI, let uri = String(data: record.type, encoding: .utf8) {
                if !hasPoster { urls.append(uri) }
            } else if let text = record.wellKnownTypeTextPayload().0 { texts.append(text) }
            else if record.typeNameFormat != .empty { unsupported.append("TNF \(record.typeNameFormat.rawValue)") }
        }
        return ["urls": urls, "texts": texts, "unsupported": unsupported]
    }
}
